gtk4/src/subclass/entry_buffer.rs

```diff
@@ -0,0 +1,326 @@
+}
+
+unsafe extern "C" fn entry_buffer_deleted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    n_chars: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let n_chars = if n_chars == u32::MAX {
+        None
+    } else {
+        Some(n_chars)
+    };
+
+    imp.deleted_text(wrap.unsafe_cast_ref(), position, n_chars)
+}
+
+static GET_TEXT_QUARK: Lazy<glib::Quark> =
+    Lazy::new(|| glib::Quark::from_string("gtk4-rs-subclass-entry-buffer-text"));
+
+unsafe extern "C" fn entry_buffer_get_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    n_bytes: *mut usize,
+) -> *const libc::c_char {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    let ret = imp.get_text(wrap.unsafe_cast_ref());
+    *n_bytes = ret.len();
+    // Ensures that the returned text stays alive for as long as
+    // the entry buffer instance
+    let fullptr = ret.to_glib_full();
+    wrap.set_qdata(*GET_TEXT_QUARK, fullptr);
+    fullptr
+}
+
+unsafe extern "C" fn entry_buffer_get_length<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+
+    imp.get_length(wrap.unsafe_cast_ref())
+}
+
+unsafe extern "C" fn entry_buffer_insert_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    charsptr: *const libc::c_char,
+    n_chars: u32,
+) -> u32 {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(charsptr);
+
+    let chars = get_text_n_chars(&text, n_chars);
+    imp.insert_text(wrap.unsafe_cast_ref(), position, chars)
+}
+
+unsafe extern "C" fn entry_buffer_inserted_text<T: EntryBufferImpl>(
+    ptr: *mut ffi::GtkEntryBuffer,
+    position: u32,
+    charsptr: *const libc::c_char,
+    length: u32,
+) {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.get_impl();
+    let wrap: Borrowed<EntryBuffer> = from_glib_borrow(ptr);
+    let text: Borrowed<GString> = from_glib_borrow(charsptr);
+
+    let chars = get_text_n_chars(&text, length);
+    imp.inserted_text(wrap.unsafe_cast_ref(), position, &chars)
+}
+
+fn get_text_n_chars(text: &str, n_chars: u32) -> &str {
+    if n_chars != u32::MAX {
+        let mut iter = text
+            .char_indices()
+            .skip(n_chars as usize)
+            .map(|(pos, _)| pos);
+        let pos_end = iter
+            .next()
+            .expect(
+                "The passed text to EntryBuffer contains fewer characters than what's passed as a length",
+            );
```

Comment:
```suggestion
            .map(|(idx, _char)| idx)
            .unwrap_or_else(|| text.len());
```

Otherwise this fails: `assert_eq!(get_text_n_chars("👨👩👧👦", 4), "👨👩👧👦");`

However that would mean that we always take the whole string if it has fewer characters too. I guess we need to skip `n_char - 1` instead and then call `next()` once to check that the character exists and another time to go behind it with `unwrap_or_else()`.

Reply:
The `n_chars - 1` did the trick

Reply:
Now you only need to also guard against `n_chars == 0`, which would wrap around now and cause problems :)

Otherwise looks good now